gh-103049: Support instruction level debugging in pdb

gaogaotiantian opened the pull request:
<!-- gh-issue-number: gh-103049 -->
* Issue: gh-103049
<!-- /gh-issue-number -->

gaogaotiantian committed:
Support assembly display on pdb

gaogaotiantian committed:
Implement ni and si, do li and lli instead of assem

artemmukhin commented on Lib/bdb.py:
```diff
@@ -272,7 +302,21 @@ def user_exception(self, frame, exc_info):
 
-    def _set_stopinfo(self, stopframe, returnframe, stoplineno=0):
+    def user_opcode(self, frame):
+        """Called when we stop or break at a opcode."""
```
```suggestion
        """Called when we stop or break at an opcode."""
```

artemmukhin commented on Lib/bdb.py:
```diff
@@ -94,6 +99,8 @@ def trace_dispatch(self, frame, event, arg):
             return self.dispatch_return(frame, arg)
         if event == 'exception':
             return self.dispatch_exception(frame, arg)
+        if event == 'opcode':
```
`opcode` event should be mentioned in "The event can be one of the following" section of `trace_dispatch` doc

artemmukhin commented on Lib/pdb.py:
```diff
@@ -330,6 +331,14 @@ def user_line(self, frame):
         if self.bp_commands(frame):
             self.interaction(frame, None)
 
+    def user_opcode(self, frame):
```
Missing docstring

artemmukhin commented on Lib/bdb.py:
```diff
@@ -309,10 +359,26 @@ def set_step(self):
+        if self.frame_returning:
+            caller_frame = self.frame_returning.f_back
+            if caller_frame and not caller_frame.f_trace:
+                caller_frame.f_trace = self.trace_dispatch
```
Consider extracting this piece into a function (e.g. `self.restore_trace_function()`) since it is used twice: in `set_step` and here

artemmukhin commented on Lib/pdb.py:
```diff
@@ -1335,29 +1351,85 @@ def do_list(self, arg):
         breaklist = self.get_file_breaks(filename)
         try:
             lines = linecache.getlines(filename, self.curframe.f_globals)
+            instructions = dis.get_instructions(self.curframe.f_code) \
```
```suggestion
            instructions = dis.get_instructions(self.curframe.f_code, adaptive=True) \
```
I am not sure, but probably we need to ask for _specialized_ bytecode that reflects the bytecode that has been executed at the moment.

artemmukhin commented on Lib/pdb.py:
```diff
@@ -1335,29 +1351,85 @@ def do_list(self, arg):
+
+        List source code for the current file with instructions.
+
+        Without arguments, list 11 lines around the current line or
```
Maybe we should mention "instructions" in this docstring, and not only "lines"

artemmukhin commented on Lib/pdb.py:
```diff
@@ -1375,13 +1447,15 @@ def do_source(self, arg):
             exc_lineno = self.tb_lineno.get(frame, -1)
         else:
             current_lineno = exc_lineno = -1
+        if instructions:
+            inst = next(instructions)
```
I suppose the whole iteration will become a simple `for instr in instructions` if we pass a list of instructions to `_print_lines` rather than an iterator. Is there any reason for passing an iterator?

gaogaotiantian replied:
Not really. We have a for loop for the lines and we need to pick the instructions that belong to this line. So we can never use a simple for loop. Another not so critical reason is that it saves some time/space because we might not exhaust the interator in `li` case.

artemmukhin replied:
Indeed, I misunderstood it at first glance. Maybe it is worth adding a comment explaining the whole line-instructions iteration

artemmukhin commented on Lib/pdb.py:
```diff
@@ -1335,29 +1351,85 @@ def do_list(self, arg):
             self.error(err)
             return
-        self._print_lines(lines, lineno, breaklist, self.curframe)
+        instructions = dis.get_instructions(self.curframe.f_code) \
```
```suggestion
        instructions = dis.get_instructions(self.curframe.f_code, adaptive=True) \
```
See the comment on `_do_list`

artemmukhin commented on Lib/pdb.py:
```diff
@@ -1559,6 +1647,13 @@ def print_stack_entry(self, frame_lineno, prompt_prefix=line_prefix):
                      self.format_stack_entry(frame_lineno, prompt_prefix))
 
+    def print_current_inst(self, frame):
+        for inst in dis.get_instructions(frame.f_code):
```
```suggestion
        for inst in dis.get_instructions(frame.f_code, adaptive=True):
```
See the comment on `_do_list`

artemmukhin reviewed (commented):
There is one more thing I would like to mention. I think it would be better to use `instr` instead of `inst` as an abbreviation for `instruction`, in both commands and source code. From what I have seen in the project, `inst` is also used for `instance`, whereas `instr` is primarily used for instruction.

gaogaotiantian committed:
Fixed a typo and use adaptive for instructions

Co-authored-by: Artem Mukhin <ortem00@gmail.com>

gaogaotiantian committed:
Fix docstrings and commands

Used stepi instead of stepinst(other commands too)

Abstracted a function to restore caller trace function

artemmukhin commented on Lib/pdb.py:
```diff
@@ -1087,6 +1099,16 @@ def do_step(self, arg):
 
+    def do_stepi(self, arg):
+        """s(tep)
+        Execute the current line, stop at the first possible occasion
```
```suggestion
        Execute the current instruction, stop at the first possible occasion
```

artemmukhin commented on Lib/bdb.py:
```diff
@@ -209,9 +234,15 @@ def stop_here(self, frame):
                 return False
-            return frame.f_lineno >= self.stoplineno
+            if self.trace_opcodes:
+                return self.lasti != frame.f_lasti
```
```suggestion
                return frame.f_lasti >= self.lasti
```
For consistency with `self.stoplineno` checks. If I understand correctly, `frame.f_lasti < self.lasti` is not supposed here (and below)

gaogaotiantian replied:
`f_lasti` could easily be going smaller (imagine a for loop). The `stoplineno` check here is for `until` command. For `next`, `self.stoplineno` would be `-1` so it's always true. The reason we needed the `f_lasti` checking here is to avoid double stop at the same instruction for both `line` and `opcode` event. We need the `line` event though to trigger breakpoints.

artemmukhin replied:
Oh, right, thanks for the explanation!

artemmukhin commented on Lib/bdb.py:
```diff
@@ -32,6 +32,9 @@ def __init__(self, skip=None):
         self.breaks = {}
         self.fncache = {}
+        self._curframe = None
+        self.lasti = -1
```
```suggestion
        self.f_lasti = -1
```
Might be more clear that it corresponds to `frame.f_lasti`

gaogaotiantian replied:
The reason I went for `lasti` instead of `f_lasti` was that's what [`dis.disassemble`](https://docs.python.org/3/library/dis.html#dis.disassemble) uses. Also in `pdb`, [`lasti2lineno`](https://github.com/python/cpython/blob/b4978ff872be5102117b4e25d93dbbb4e04c8292/Lib/pdb.py#L110) uses `lasti`. The `f` in `f_lasti` is a prefix for frame attributes, and I think `lasti` is a reasonable choice here.

artemmukhin reviewed (commented):
Using `stepi` and `nexti` as long versions of commands makes sense for me as well 👍

gaogaotiantian committed:
Fix docstring for stepi

Co-authored-by: Artem Mukhin <artem.m.mukhin@gmail.com>

gaogaotiantian committed:
Add tests for instruction level commands

gaogaotiantian committed:
Add docs for instruction commands

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

artemmukhin commented on Lib/pdb.py:
```diff
@@ -1096,6 +1118,15 @@ def do_next(self, arg):
+    def do_nexti(self, arg):
+        """n(ext)
+        Continue execution until the next line in the current function
+        is reached or it returns.
```
Needs to be updated

artemmukhin commented on Doc/library/pdb.rst:
```diff
@@ -379,6 +379,13 @@ can be overridden by the local file.
 
+.. pdbcommand:: si | stepi
+
+   Execute the current instruction, stop at the first possible occasion (either in a
```
```suggestion
   Execute the current bytecode instruction, stop at the first possible occasion (either in a
```

I guess it might be unclear what "instruction" mean. The same applies to other mentions of "instruction".

artemmukhin commented on Lib/pdb.py:
```diff
@@ -1335,29 +1352,87 @@ def do_list(self, arg):
+    def do_listi(self, arg):
+        """listi | li [first[, last] | .]
+
+        List source code for the current file with instructions.
```
```suggestion
        List source code for the current file with bytecode instructions.
```
The same as in `pdb.rst`

artemmukhin commented on Lib/test/test_pdb.py:
```diff
@@ -2351,6 +2352,101 @@ def _create_fake_frozen_module():
+        curr_instr_lines = [line.strip() for line in stdout.splitlines() if "-->" in line]
+        self.assertEqual(len(curr_instr_lines), 3)
+        for line in curr_instr_lines:
+            # Make sure ni is moving forward, not stopping at the same instrunction
```
```suggestion
            # Make sure ni is moving forward, not stopping at the same instruction
```

gaogaotiantian committed:
Fixed a typo in comment

Co-authored-by: Artem Mukhin <artem.m.mukhin@gmail.com>

gaogaotiantian committed:
Update some docs and comments
